### PR TITLE
Issue #2828525: Implement manual payment gateways

### PR DESCRIPTION
**Pull request by vasike**

### CODE REVIEW

File: modules/payment/src/Plugin/Commerce/PaymentGateway/Manual.php [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    if (!$form_state->getErrors()) {
+      $form_parents = $form['#parents'];
+      $form_parents[] = 'manual';
+      $values = $form_state->getValue($form_parents);
+      if (!empty($values['expires'])) {
+        $convert = strtotime($values['expires']);
+        if ($convert == -1 || $convert === FALSE) {
+          $form_state->setError($form['manual']['expires'], $this->t('Invalid offset time format.'));
+        }
+        if ($convert < REQUEST_TIME) {
```
Comment (mglaman) [inline modules/payment/src/Plugin/Commerce/PaymentGateway/Manual.php]: :) Darn `REQUEST_TIME` global

File: modules/payment/src/Plugin/Commerce/PaymentGateway/Manual.php [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    $payment_method = $payment->getPaymentMethod();
+    if (empty($payment_method)) {
+      throw new \InvalidArgumentException('The provided payment has no payment method referenced.');
+    }
+    if ($payment_method->isExpired()) {
+      throw new HardDeclineException('The provided payment method has expired');
+    }
+
+    $payment->state = 'pending';
+    $payment->setAuthorizedTime(REQUEST_TIME);
```
Comment (mglaman) [inline modules/payment/src/Plugin/Commerce/PaymentGateway/Manual.php]: Deprecated global usage

**Review by mglaman — CHANGES_REQUESTED** [review]
Various `REQUEST_TIME` usages, and I do not see any tests

File: modules/payment/src/Plugin/Commerce/PaymentGateway/Manual.php (hunk 2 above)
Comment (vasike) [reply modules/payment/src/Plugin/Commerce/PaymentGateway/Manual.php]: i was aware. i'll replace them

File: modules/payment/tests/src/Functional/ManualPaymentAdminTest.php [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+    $this->order = $this->createEntity('commerce_order', [
+      'uid' => $this->loggedInUser->id(),
+      'type' => 'default',
+      'state' => 'draft',
+      'order_items' => [$order_item],
+      'store_id' => $this->store,
+    ]);
+
+    $this->paymentUri = 'admin/commerce/orders/' . $this->order->id() . '/payments';
```
Comment (mglaman) [inline modules/payment/tests/src/Functional/ManualPaymentAdminTest.php]: We should generate this via `Url::fromRoute` with proper parameters. I know `drupalGet` tries, but we should use proper generation.

**Review by mglaman — CHANGES_REQUESTED** [review]
Anything manually building a path for `drupalGet` should use `Url::fromRoute`.

Also, I'd like an "admin" test which verifies the creation of a manual payment method from the config area.

**Comment by bojanz** [comment]
Final version committed in 753eabf8db02e1d2a9e167056b5e34bf5a763aa1.

**Pull request closed without merging**